src/Traverser.php

```diff
@@ -116,60 +117,54 @@ protected function getChildrenFromParent(): Collection
 
     public function ancestors(): Collection
     {
-        return $this->getAncestors(
-            $this->getDocuments()->get($this->id)
-        );
+        return $this->getAncestors()->map(function($id) {
+            return $this->query->find($id);
+        });
     }
 
     public function ancestorsAndSelf(): Collection
     {
-        return $this->ancestors()->push(
-            $this->getDocuments()->get($this->id)
-        );
+        return $this->ancestors()->push($this->document);
     }
 
-    protected function getAncestors(Document $document, $ancestors = null): Collection
+    protected function getAncestors(): Collection
     {
-        if ( ! $ancestors instanceof Collection) {
-            $ancestors = collect();
-        }
+        $ancestors = collect();
 
-        $parentMethod = $this->getParentMethod($document);
+        $parentMethod = $this->getParentMethod($this->document);
 
-        if ($parent = $document->{$parentMethod}()) {
-            $ancestors->prepend($parent);
-            $this->getAncestors($parent, $ancestors);
+        if ($parent = $this->document->{$parentMethod}()) {
+            $ancestors->prepend($parent->id);
+
+            $ancestors = $parent->traverse()->getAncestors()->merge($ancestors);
```

Comment:
I don't think this allows for different instances of the the Traverse class to be used as it tight-couples it with the 'default' `traverse` method on the Document.

Reply:
Agreed, this should instantiate the Traverse class manually for recursion like this.